File: infra/lambda/test/entity/hero.test.ts
import { HeroSchema } from "../../entity/hero";

const createHero = ()=>({
    createdAt: '2026-01-01',
    name: 'Invoker',
    picks: 10,
    wins: 6,
    tier: 'S',
    winRate: 60
});
test('HeroSchema 正常データ', () => {

    const parsed = HeroSchema.parse(createHero());

    expect(parsed).toEqual(createHero());
});

test('名前が空の時エラーを返すこと', () => {
    const heroWithInvalidName = {
        ...createHero(),
        name:'',
    }

    expect(() => HeroSchema.parse(heroWithInvalidName))
        .toThrow();
});

test('ピック数、勝利数が０未満の時にエラーを返すこと',()=>{
    const heroWithInvalidPicks = {
        ...createHero(),
        picks:-1
    };
    const heroWithInvalidWins = {
        ...createHero(),
        wins:-1
    };

    expect(()=>HeroSchema.parse(heroWithInvalidPicks))
    .toThrow();
    expect(()=>HeroSchema.parse(heroWithInvalidWins))
    .toThrow();
    
});

test('ティアがS,A,B,Cのいずれかでない時にエラーを返すこと',()=>{
    const heroWithInvalidTier = {
        ...createHero(),
        tier:'D'
    };

    expect(()=>HeroSchema.parse(heroWithInvalidTier))
    .toThrow();
    
});

test('勝率が不適切な場合はエラーを返すこと',()=>{
    const heroWithInvalidWinRate1 = {
        ...createHero(),
        winRate:-1
    };
    const heroWithInvalidWinRate2 = {
        ...createHero(),
        winRate:120
    };

    expect(()=>HeroSchema.parse(heroWithInvalidWinRate1))
    .toThrow();
    expect(()=>HeroSchema.parse(heroWithInvalidWinRate2))
    .toThrow();
    
});

File: frontend/test/features/hero/domain/entity/hero_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:frontend/features/hero/domain/entity/hero.dart';

void main(){
  group('Hero.create (Entity Validation)', (){
    final validDate = '2036-03-03';

    test('正常な入力でインスタンスが生成されること',(){
      final hero = Hero.create(
        createdAt: validDate, 
        name: 'Mars', 
        picks: 100, 
        wins: 60, 
        tier: Tier.S, 
        winRate: 60);

      expect(hero.name, 'Mars');
    });
    test('名前が空の時にエラーを投げること', (){
        expect(()=>Hero.create(
        createdAt: validDate, 
        name: '', 
        picks: 100, 
        wins: 60, 
        tier: Tier.S, 
        winRate: 60), throwsA(isA<ArgumentError>()));
    });
    test('ピック数または勝利数が負の時にエラーを投げること', (){
        expect(()=>Hero.create(
        createdAt: validDate, 
        name: 'Mars', 
        picks: -1, 
        wins: 60, 
        tier: Tier.S, 
        winRate: 60), throwsA(isA<ArgumentError>()));

        expect(()=>Hero.create(
        createdAt: validDate, 
        name: 'Mars', 
        picks: 100, 
        wins: -1, 
        tier: Tier.S, 
        winRate: 60), throwsA(isA<ArgumentError>()));
    });
    test('勝率が0未満または100より大きい時にエラーを投げること', (){
        expect(()=>Hero.create(
        createdAt: validDate, 
        name: 'Mars', 
        picks: 100, 
        wins: 60, 
        tier: Tier.S, 
        winRate: 101), throwsA(isA<ArgumentError>()));

        expect(()=>Hero.create(
        createdAt: validDate, 
        name: 'Mars', 
        picks: 100, 
        wins: 60, 
        tier: Tier.S, 
        winRate: -1), throwsA(isA<ArgumentError>()));
    });
  });
}